Support JDK9 methods using MR-JAR

FroMage opened the pull request:
WDYT @manovotn ? I'm not too sure how to test this though. I have to check how other projects do it in the CI.

manovotn commented on core/pom.xml:
```diff
@@ -9,6 +10,71 @@
+                <artifactId>maven-antrun-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>compile-java9</id>
```
Shouldn't we now enforce min JDK version to build the project?

FroMage replied:
Ah, good point.

manovotn commented on tests/src/test/java/io/smallrye/context/test/MiscTest.java:
```diff
@@ -0,0 +1,23 @@
+public class MiscTest {
+
+    @Test
+    public void testCFWrapping() {
```
What does this test exactly? Isn't the assert always true?

manovotn replied:
Not sure I follow, can you expand please?

manovotn commented on core/pom.xml:
```diff
@@ -9,6 +10,71 @@
+                            <tasks>
+                                <property name="compile_classpath" refid="maven.compile.classpath" />
+                                <mkdir dir="${java9.build.outputDirectory}" />
+                                <javac fork="true" executable="${env.JAVA_HOME}/bin/javac" srcdir="${java9.sourceDirectory}"
```
Also relevant to my previous comment, assuming we will enforce building on JDK 9+ (most likely 11), then you shouldn't need this. Having to have env variable defined to make a build pass is sort of weird, ideally it should all "just work" with `mvn clean install` ;-)

FroMage commented:
@manovotn I've updated the way I do the MR to the first option recommended on https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html which means we have two virtual modules:

- `core-java8` (the base)
- `core-java9` (the MR for 9)

and `core` just assembles the two into an MR jar.

It's more standard build for Maven I guess, but it has two drawbacks:

- Eclipse can't load the modules properly because it doesn't see that the `core` module has classes, so you need to add the `core-java8/9` projects to its build path.
- Dependencies from `core-java8` need to be duplicated in `core` because that's the pom that will be used.

I added tests for the new methods too.

FroMage commented:
Updated it again with the structure supported by smallrye-parent, that's much cleaner.

Ladicek commented on core/bin/mrjar.xml:
```diff
@@ -0,0 +1,37 @@
+      </binaries>
+    </moduleSet>
+  </moduleSets>
+</assembly>
```
Is this file still used?

manovotn replied:
This should be safe to remove.

FroMage replied:
Ah yes, thanks.

manovotn commented on core/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1,12 @@
+          http://java.sun.com/xml/ns/javaee http://docs.jboss.org/cdi/beans_1_0.xsd
+          http://jboss.org/schema/weld/beans http://jboss.org/schema/weld/beans_1_1.xsd">
+    <weld:scan>
+        <!-- Don't let Arquillian choke on MR-Jars -->
```
Hmm this should be fixed in more recent version of the container - https://github.com/arquillian/arquillian-container-weld/commit/7296f97caac5ceaf8a5ec3a11a483a7c36e84a44#diff-797455798ac6dee103ff0fd610638bb5
Version of arq container seems to come from SR parent and newer versions are already updated, just bump it and you can remove this file :)

manovotn replied:
Actually there is already https://github.com/smallrye/smallrye-context-propagation/pull/181
So you can choose between merging and rebasing or updating it as part of this PR, up to you.

FroMage replied:
Well spotted, I've merged #181, rebased and removed that file.

manovotn commented on tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            CompletableFuture.class.getDeclaredMethod("copy");
+        } catch (NoSuchMethodException | SecurityException e) {
+            Assume.assumeTrue(false);
```
I know it's nitpicking but why not just `return`?
Also, a comment explaining why this is ok might be good.

FroMage replied:
If you return the test is not marked as skipped.

FroMage replied:
I added a comment.

manovotn replied:
> If you return the test is not marked as skipped.

Ah, fair enough, didn't realize that.

FroMage commented:
I think this PR's CI is toast, lemme reopen one.

FroMage commented:
Finally, CI rules are fixed.

dmlloyd commented:
> I am probably missing something but why do you need empty files such as `core/build-release-8` or `tests/build-release-8`?
> 
> Otherwise looks pretty good.

Have a look at the docs here: https://github.com/smallrye/smallrye-parent/blob/master/README.adoc
